File: py/candlestick_patterns/patterns/thrusting.py
"""Thrusting pattern (2-candle bearish)."""
from __future__ import annotations

from ..core.primitives import is_white, is_black, real_body


def thrusting(self) -> int:
    """Thrusting: a two-candle bearish continuation pattern.

    Must have:
    - first candle: long black,
    - second candle: white, opens below the prior candle's low, closes
      into the prior candle's real body but below the midpoint, and the
      close is not equal to the prior candle's close (to distinguish
      from in-neck).

    The meaning of "long" is specified with self._long_body.
    The meaning of "equal" is specified with self._equal.

    Returns:
        -100 for bearish, 0 for no pattern.
    """
    if not self._enough(2, self._long_body, self._equal):
        return 0

    o1, h1, l1, c1 = self._bar(2)
    o2, h2, l2, c2 = self._bar(1)

    rb1 = real_body(o1, c1)
    eq = self._avg(self._equal, 2)

    if (is_black(o1, c1) and is_white(o2, c2) and
            rb1 > self._avg(self._long_body, 2) and
            o2 < l1 and
            c2 > c1 + eq and
            c2 <= c1 + rb1 * 0.5):
        return -100

    return 0

File: ts/candlestick-patterns/patterns/thrusting.ts
/** Thrusting pattern (2-candle bearish). */
import { CandlestickPatternsEngine } from '../core/engine.ts';
import { tProductAll } from '../../fuzzy/index.ts';
import { realBody, isBlack, isWhite } from '../core/primitives.ts';

/**
 * Thrusting: a two-candle bearish continuation pattern.
 *
 * Must have:
 * - first candle: long black,
 * - second candle: white, opens below the prior candle's low, closes
 *   into the prior candle's real body but below the midpoint, and the
 *   close is not equal to the prior candle's close (to distinguish
 *   from in-neck).
 *
 * The meaning of "long" is specified with `longBody`.
 * The meaning of "equal" is specified with `equal`.
 *
 * Category A: always bearish (continuous).
 *
 * Returns:
 *     Continuous float in [-100, 0].  Always bearish.
 */
export function thrusting(cp: CandlestickPatternsEngine): number {
    if (!cp.enough(2, cp.longBody, cp.equal)) return 0.0;

    const b1 = cp.bar(2);
    const b2 = cp.bar(1);

    const rb1 = realBody(b1.o, b1.c);

    // Crisp gates: color checks and open below prior low.
    if (!(isBlack(b1.o, b1.c) && isWhite(b2.o, b2.c) && b2.o < b1.l)) return 0.0;

    // Fuzzy conditions.
    const muLong1 = cp.muGreaterCS(rb1, cp.longBody, 2);

    // Close above prior close + equal avg (not equal to prior close).
    const eq = cp.avgCS(cp.equal, 2);
    let eqWidth = 0.0;
    if (eq > 0.0) eqWidth = cp.fuzzRatio * eq;
    const muAboveClose = cp.muGtRaw(b2.c, b1.c + eq, eqWidth);

    // Close at or below midpoint of prior body: c2 <= c1 + rb1 * 0.5
    const mid = b1.c + rb1 * 0.5;
    let midWidth = 0.0;
    if (rb1 > 0.0) midWidth = cp.fuzzRatio * rb1 * 0.5;
    const muBelowMid = cp.muLtRaw(b2.c, mid, midWidth);

    return -tProductAll(muLong1, muAboveClose, muBelowMid) * 100.0;
}
